### DISCUSSION TITLE
What are these extended filenames with tilde (~)?

### DISCUSSION THREAD

**Discussion by tom-andersson** [abraunegg/onedrive#1763] [Q&A] [answered]
When I `watch` the output of `systemctl --user status onedrive` and I save a file tracked by `onedrive`, I see the following logs from the systemctl service:
```
Dec 30 15:00:38 bsl-tomand-l1 onedrive[2107]: Cannot upload file changes/creation: vimwiki/4913: No such file or directory
Dec 30 15:00:38 bsl-tomand-l1 onedrive[2107]: Moving vimwiki/index.wiki to vimwiki/index.wiki~
Dec 30 15:00:38 bsl-tomand-l1 onedrive[2107]: Cannot move item: vimwiki/index.wiki~: No such file or directory
```
What is the reason for the attempt to rename the file with the filename extended with a `~` character? And secondly, why is the service attempting to upload a file (that doesn't exist) called `4913` in the folder of the file I'm saving? I'd be grateful to understand these things because I'm unsure if I should be worried about these logs.

**Comment by abraunegg** [comment, marked as answer]
@tom-andersson 
> `Dec 30 15:00:38 bsl-tomand-l1 onedrive[2107]: Cannot upload file changes/creation: vimwiki/4913: No such file or directory`
> ...
> And secondly, why is the service attempting to upload a file (that doesn't exist) called 4913 in the folder of the file I'm saving?

This is because you are editing files using 'vim' .. pick a different editor that does not have this method of saving files. A Google search will help you understand the cause in greater detail.

> What is the reason for the attempt to rename the file with the filename extended with a ~ character?

This is possibly linked to the above.

When you are editing the file, the editor is using '4913' initially, then using the filename with a ~ at the end before writing the file out correctly.

Nothing to worry about here. If you are 'worried' you can exclude '4913' and files ending in '~' to prevent this situation from occurring. 

Modify your 'config' may assist:
```
skip_file = "~*|.~*|*.tmp|4913|*~"
```

If you do make this sort of change, test by using `--dry-run` to confirm this is performing as required.

**Reply by tom-andersson** [reply]
Thank you for the help @abraunegg. Sorry for not realising this was actually a Vim issue. I took these steps to fix the problem:
  - Added `*~|4913$` to `skip_file` in my config file. The dollar sign in `4913$` was from a previous attempt where I mistakenly used the regex-style character for 'end of string'. But I believe it is actually interpreted as a wildcard for '0 or one character'. Is that correct? Thus, `4913` will still correctly be skipped.
  - Closed all onedrive processes
  - Ran `onedrive --resync --synchronize`. Thankfully it didn't result in a full database refresh - perhaps because of the minimal change to the config. Or because it didn't detect any existing local files that matched the new expressions (because they are only temporary files during writing in Vim). That being said, I'm unsure if that's the correct interpretation and a bit confused as to why the resync only took a few seconds rather than five minutes as it had previously. The output of the command is as below:
```
Configuration file successfully loaded
Deleting the saved application sync status ...
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Uploading differences of ~/OneDrive
Uploading new items of ~/OneDrive
```
  - I restarted the onedrive `systemctl` service, I opened a file tracked by onedrive in vim and saved a change with `:w`. I confirmed from the `journalctl` logs that uploads of the temporary files were not attempted.

**Reply by m-guggenmos** [reply]
@abraunegg I just stumbled across this issue. In my case, adding `*~` to `skip_file` results in a deletion of the main file on OneDrive. For instance, editing and saving the file `text.txt` with vim results in the temporary creation of the file `text.txt~`. Including `*~ to` in the `skip_file` option successfully prevents an upload of the file `text.txt~`. However, it also deletes the `text.txt` file itself from OneDrive! (`Deleting item from OneDrive: test.txt`). It is uploaded again with the next refresh.

Not including `*~` in `skip_file` resolves this, but then of course onedrive tries to upload `text.txt~` (which also fails, possibly because it's so shortlived).

This is my `skipe_file` line in the config file:
```
skip_file = "~*|.~*|*.tmp|*.swp|*.swx|4913|#*|*~"
```

**Reply by abraunegg** [reply]
@m-guggenmos 
> Not including *~ in skip_file resolves this, but then of course onedrive tries to upload text.txt~ (which also fails, possibly because it's so shortlived).

Use:  `skip_file = "~*|.~*|*.tmp|*.swp|*.swx|4913|#*|.txt~"` or `skip_file = "~*|.~*|*.tmp|*.swp|*.swx|4913|#*|.*~"`

**Reply by km22** [reply]
Because the temporary files are so shortlived, is there an option to incorporate a slight delay in the sync such that if the temporary files exist less than, say, 25ms then they are simply ignored?

**Reply by abraunegg** [reply]
@km22 
The application already 'checks' for short lived files that trigger an upload request, then when attempted to be uploaded and no longer exist, are skipped as they cannot be uploaded.

When this occurs, the following message is logged: `Skipping item - path has disappeared: <path>`
